### PR TITLE
[WIP] Allow setting threshold for diff in order to use via CI

### PR DESCRIPTION
**Pull request by liiight**
Setting exit status via predefined threshold will allow using wily via CI. Relating to issue #66.

# Changes 

* Added ability to pass thresholds to diff command per operator

# Usage:

```bash
$ wily diff [FILES] --thresholds halstead.h1=1,raw.loc=50
```

## TODO

- [x] Check for metric type when determining if threshold was passed or not
- [ ] Handle diffs for scenarios where files/modules were renamed, resulting in `-` as the base value
- [x] Add ability to pass threshold via config file
- [ ] Add more tests
- [ ] Have a nicer output for errors
- [ ] Add logging

### CODE REVIEW

**Comment by liiight** [comment]
So I've been thinking about the use case of new or renamed modules. The way that it's currently implemented is by calculating deltas, which obviously does not make sense for this use case. But what if instead, or possibly in addition to deltas, we'll allow setting maximal/minimal values as thresholds?

WDYT?

**Comment by liiight** [comment]
@tonybaloney I'd love your feedback on how best to handle the renamed/new modules use case. Also, I'm not sure why that test fail in the CI

File: wily/__main__.py [hunk 1]
```diff
@@ -223,8 +239,14 @@ def report(ctx, file, metrics, number, message, format, console_format, output):
     default=True,
     help="Show function/class level metrics where available",
 )
+@click.option(
+    "--thresholds",
+    type=ThresholdsParamType(),
```
Comment (tonybaloney) [inline wily/__main__.py]: Are you sure you pass an instance here and not a type?

**Comment by tonybaloney** [comment]
I think the threshold parameters should be a multiple argument, instead of a CSV,

So:

```
wily diff --threshold h1=2 --threshold loc=200
```

Instead of:

```
wily diff --threshold h1=2,loc=200
```

Users would be less likely to mistakenly add a space in the list, also it's easier to explain in documentation (and requires less code).

Click supports this out of the box.

**Comment by tonybaloney** [comment]
The test is failing because it's not picking up an existing cache, so it's prompting to rebuild a new one.

Try repro'ing it locally on Windows if you have access to windows.

Also, look at the fixtures for a working git repo used by the other tests.

File: wily/__main__.py (hunk 1 above)
Comment (liiight) [reply wily/__main__.py]: [yep](https://click.palletsprojects.com/en/7.x/parameters/#implementing-custom-types). I also added a test for this.

**Comment by liiight** [comment]
> I think the threshold parameters should be a multiple argument
That would be my first choice as well, but I was consistent with how metrics are passed. I can change that too if you'd like (could also make that a custom type as well).

There are other place that could use some beefing up using click options, not sure if this is the best place to discuss those though.

> Also, look at the fixtures for a working git repo used by the other tests.
I have and I already used it in other tests that I added. For some reason this is the only test that fails for me (works when running it locally though).

BTW: Maybe it would be beneficial making `gitdir` and `builddir` into session scoped fixtures?

**Comment by liiight** [comment]
@tonybaloney I plan to switch to a slightly different approach, have the ability to set delta threshold and absolute threshold. This will allow handling new/renamed files and also could prove to be more useful as well.

**Comment by tonybaloney** [comment]
@liiight are you still interested in working on this?

**Pull request closed without merging**